pkg/stats/monitor.go

```diff
@@ -171,45 +188,80 @@ func (m *Monitor) CanAcceptRequest(req *rpc.StartEgressRequest) bool {
+func (m *Monitor) AcceptRequest(req *rpc.StartEgressRequest) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if !m.CanAcceptRequest(req) {
+		return errors.ErrResourceExhausted
+	}
+
 	var cpuHold float64
 	switch req.Request.(type) {
 	case *rpc.StartEgressRequest_RoomComposite:
-		cpuHold = m.cpuCostConfig.RoomCompositeCpuCost * 1.2
+		m.reserved += m.cpuCostConfig.RoomCompositeCpuCost
+		cpuHold = m.cpuCostConfig.RoomCompositeCpuCost
 	case *rpc.StartEgressRequest_Web:
-		cpuHold = m.cpuCostConfig.WebCpuCost * 1.2
+		m.reserved += m.cpuCostConfig.WebCpuCost
+		cpuHold = m.cpuCostConfig.WebCpuCost
 	case *rpc.StartEgressRequest_TrackComposite:
-		cpuHold = m.cpuCostConfig.TrackCompositeCpuCost * 1.2
+		m.reserved += m.cpuCostConfig.TrackCompositeCpuCost
+		cpuHold = m.cpuCostConfig.TrackCompositeCpuCost
 	case *rpc.StartEgressRequest_Track:
-		cpuHold = m.cpuCostConfig.TrackCpuCost * 1.2
+		m.reserved += m.cpuCostConfig.TrackCpuCost
+		cpuHold = m.cpuCostConfig.TrackCpuCost
 	}
 
 	m.pendingCPUs.Add(cpuHold)
 	time.AfterFunc(cpuHoldDuration, func() { m.pendingCPUs.Sub(cpuHold) })
+	return nil
 }
 
 func (m *Monitor) EgressStarted(req *rpc.StartEgressRequest) {
 	switch req.Request.(type) {
 	case *rpc.StartEgressRequest_RoomComposite:
-		m.requestGauge.With(prometheus.Labels{"type": "room_composite"}).Add(1)
+		m.requestGauge.With(prometheus.Labels{"type": roomComposite}).Add(1)
 	case *rpc.StartEgressRequest_Web:
-		m.requestGauge.With(prometheus.Labels{"type": "web"}).Add(1)
+		m.requestGauge.With(prometheus.Labels{"type": web}).Add(1)
 	case *rpc.StartEgressRequest_TrackComposite:
-		m.requestGauge.With(prometheus.Labels{"type": "track_composite"}).Add(1)
+		m.requestGauge.With(prometheus.Labels{"type": trackComposite}).Add(1)
 	case *rpc.StartEgressRequest_Track:
-		m.requestGauge.With(prometheus.Labels{"type": "track"}).Add(1)
+		m.requestGauge.With(prometheus.Labels{"type": track}).Add(1)
 	}
 }
 
 func (m *Monitor) EgressEnded(req *rpc.StartEgressRequest) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	switch req.Request.(type) {
+	case *rpc.StartEgressRequest_RoomComposite:
+		m.reserved -= m.cpuCostConfig.RoomCompositeCpuCost
+		m.requestGauge.With(prometheus.Labels{"type": roomComposite}).Sub(1)
+	case *rpc.StartEgressRequest_Web:
+		m.reserved -= m.cpuCostConfig.WebCpuCost
+		m.requestGauge.With(prometheus.Labels{"type": web}).Sub(1)
+	case *rpc.StartEgressRequest_TrackComposite:
+		m.reserved -= m.cpuCostConfig.TrackCompositeCpuCost
+		m.requestGauge.With(prometheus.Labels{"type": trackComposite}).Sub(1)
+	case *rpc.StartEgressRequest_Track:
+		m.reserved -= m.cpuCostConfig.TrackCpuCost
+		m.requestGauge.With(prometheus.Labels{"type": track}).Sub(1)
+	}
+}
+
+func (m *Monitor) EgressAborted(req *rpc.StartEgressRequest) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
 	switch req.Request.(type) {
 	case *rpc.StartEgressRequest_RoomComposite:
-		m.requestGauge.With(prometheus.Labels{"type": "room_composite"}).Sub(1)
```

Comment:
we don't need to subtract from gauge?

Reply:
this is called when something fails when EgressStarted hasn't been called yet, but the request had already been accepted